src/tactic/nth_rewrite/default.lean

```diff
@@ -0,0 +1,154 @@
+* `nth_rewrite_rhs`: as above, but only rewrites on the right hand side of an equation or iff.
+
+## Implementation details
+
+There are two alternative backends, provided by `.congr` and `.kabstract`.
+The kabstract backend is not currently available through mathlib.
+
+The kabstract backend is faster, but if there are multiple identical occurences of the
+same rewritable subexpression, all are rewritten simultaneously,
+and this isn't always what we want.
+(In particular, `rewrite_search` is much less capable on the `category_theory` library.)
+-/
+
+open tactic lean.parser interactive interactive.types
+
+namespace tactic
+
+/-- Returns the target of the goal when passed `none`,
+otherwise, return the type of `h` in `some h`. -/
+meta def target_or_hyp_type : option expr → tactic expr
+| none     := target
+| (some h) := infer_type h
+
+namespace nth_rewrite
+
+/-- Wrapper that will either replace the target, or a hypothesis,
+depending on whether `none` or `some h` is given as the first argument. -/
+meta def replace' : option expr → expr → expr → tactic unit
+| none     := tactic.replace_target
+| (some h) := λ e p, tactic.replace_hyp h e p >> skip
+
+/-- A helper function for building the proof witness of a rewrite
+on one side of an equation of iff. -/
+meta def mk_lambda (r lhs rhs : expr) : side → tactic expr
+| side.L := do L ← infer_type lhs >>= mk_local_def `L, lambdas [L] (r L rhs)
+| side.R := do R ← infer_type rhs >>= mk_local_def `R, lambdas [R] (r lhs R)
+
+/-- A helper function for building the new total expression
+starting from a rewrite of one side of an equation or iff. -/
+meta def new_exp (exp r lhs rhs : expr) : side → expr
+| side.L := r exp rhs
+| side.R := r lhs exp
+
+/-- Given a tracked rewrite of (optionally, a side of) the target or a hypothesis,
+update the tactic state by replacing the corresponding part of the tactic state
+with the rewritten expression. -/
+meta def replace : option side → option expr → tracked_rewrite → tactic unit
+| none     := λ h rw, do (exp, prf) ← rw.eval, replace' h exp prf
+| (some s) := λ h rw,
+  do (exp, prf) ← rw.eval,
+     expr.app (expr.app r lhs) rhs ← target_or_hyp_type h,
+     lam ← mk_lambda r lhs rhs s,
+     new_prf ← mk_congr_arg lam prf,
+     replace' h (new_exp exp r lhs rhs s) new_prf
+
+end nth_rewrite
+
+open nth_rewrite nth_rewrite.congr nth_rewrite.tracked_rewrite
+open tactic.interactive
+
+/-- Preprocess a rewrite rule for use in `get_nth_rewrite`. -/
+private meta def unpack_rule (p : rw_rule) : tactic (expr × bool) :=
+do r ← to_expr p.rule tt ff,
+   return (r, p.symm)
+
+/-- Get the `n`th rewrite of rewrite rules `q` in expression `e`,
+or fail if there are not enough such rewrites. -/
+private meta def get_nth_rewrite (n : ℕ) (q : rw_rules_t) (e : expr) :
+  tactic tracked_rewrite :=
+do rewrites ← q.rules.mmap $ λ r, unpack_rule r >>= nth_rewrite e,
+   rewrites.join.nth n <|> fail format!"failed: not enough rewrites found"
```

Comment:
Unnecessary `format!`?

Reply:
If you say so... there is still a lot of the tactic framework that is a mystery to me.
Thanks for pointing this out!

Reply:
Huh? So why did you put the `format!` back?

Reply:
Sorry, it was an accident!